src/client/datascience/jupyter/jupyterCommand.ts

```
@@ -0,0 +1,159 @@
+    private interpreterPromise: Promise<PythonInterpreter | undefined>;
+    private activationHelper: IEnvironmentActivationService;
+
+    constructor(exe: string, args: string[], processServiceFactory: IProcessServiceFactory, activationHelper: IEnvironmentActivationService, interpreterService: IInterpreterService) {
+        this.exe = exe;
+        this.requiredArgs = args;
+        this.launcherPromise = processServiceFactory.create();
+        this.activationHelper = activationHelper;
+        this.interpreterPromise = interpreterService.getInterpreterDetails(this.exe).catch(e => undefined);
+    }
+
+    public async mainVersion(): Promise<number> {
+        const interpreter = await this.interpreterPromise;
+        if (interpreter && interpreter.version) {
+            return interpreter.version.major;
+        } else {
+            return this.execVersion();
+        }
+    }
+
+    public interpreter() : Promise<PythonInterpreter | undefined> {
+        return this.interpreterPromise;
+    }
+
+    public async execObservable(args: string[], options: SpawnOptions): Promise<ObservableExecutionResult<string>> {
+        const newOptions = { ...options };
+        newOptions.env = await this.fixupEnv(newOptions.env);
+        const launcher = await this.launcherPromise;
+        const newArgs = [...this.requiredArgs, ...args];
+        return launcher.execObservable(this.exe, newArgs, newOptions);
+    }
+
+    public async exec(args: string[], options: SpawnOptions): Promise<ExecutionResult<string>> {
+        const newOptions = { ...options };
+        newOptions.env = await this.fixupEnv(newOptions.env);
+        const launcher = await this.launcherPromise;
+        const newArgs = [...this.requiredArgs, ...args];
+        return launcher.exec(this.exe, newArgs, newOptions);
+    }
+
+    private async execVersion(): Promise<number> {
+        const launcher = await this.launcherPromise;
+        if (launcher) {

```

exec and execObservable both just await this.launcherPromise and then directly use the launcher. Do you need this check here? If you do I assume it would be needed for the other two execs.

---

If launcher isn't found for the other two, it should throw an exception. 

---
In reply to: [248088450](https://github.com/Microsoft/vscode-python/pull/3998#discussion_r248088450) [](ancestors = 248088450)